### PR TITLE
Implement unpacking nested pytest.param objects in parametrize

### PR DESCRIPTION
**Pull request by vgorkavenko**
This PR result of the discussion in https://github.com/pytest-dev/pytest/issues/9074

Now nested `pytest.param` objects will be unpacked to parent: nested values are placed in parent `pytest.param`, marks are merged, id is merged if parent `pytest.param` does not have id

- [ ] Include documentation when adding new features.
- [X] Include new tests or update existing tests when applicable.
- [X] Allow maintainers to push and squash when merging my commits. Please uncheck this if you prefer to squash the commits yourself.
- [X] Create a new changelog file in the `changelog` folder, with a name like `<ISSUE NUMBER>.<TYPE>.rst`. See 
- [X] Add yourself to `AUTHORS` in alphabetical order.

### CODE REVIEW

**Review by RonnyPfannschmidt — COMMENTED** [review]
I like the initial idea as it enhances composition and user experience

Strictly speaking it may be necessary to start this as a opt out,

But enabling a opt out may be enough

Depending on luck i may be able to give this a deeper look later today

**Comment by vgorkavenko** [comment]
@RonnyPfannschmidt, can we invite someone else to review?

File: changelog/9074.improvement.rst [hunk 1]
```diff
@@ -0,0 +1,17 @@
+Now ``pytest`` do unpack nested ``pytest.param`` values and merge their ids and marks.
```
Comment (nicoddemus) [inline changelog/9074.improvement.rst:1]:
```suggestion
Now ``pytest`` unpacks nested ``pytest.param`` values and merges their ids and marks.
```

File: changelog/9074.improvement.rst [hunk 2]
```diff
@@ -0,0 +1,17 @@
...
+If parent ``pytest.param`` has an id - it rewrites nested ids.
```
Comment (nicoddemus) [inline changelog/9074.improvement.rst:2]: I think we can leave this phrase out, seems a bit too detailed for the changelog (this should live in its own docs).

```suggestion

```

File: changelog/9074.improvement.rst [hunk 3]
```diff
@@ -0,0 +1,17 @@
...
+                    pytest.param(1, id="one", marks=pytest.mark.one),
+                    pytest.param(2, id="two", marks=pytest.mark.two),
+                    pytest.param(3, id="three", marks=pytest.mark.three),
+                    marks=pytest.mark.full
+                )
+            ]
+        )
+
+
+Earlier in this case the values ``a, b, c`` in test are a ``ParameterSet`` objects
```
Comment (nicoddemus) [inline changelog/9074.improvement.rst:17]:
```suggestion
In previous versions, ``a, b, c`` were passed as ``ParameterSet`` objects.
```

File: src/_pytest/python.py [hunk 4]
```diff
@@ -1126,9 +1126,23 @@ def parametrize(
         # Create the new calls: if we are parametrize() multiple times (by applying the decorator
         # more than once) then we accumulate those calls generating the cartesian product
         # of all calls.
+        # If we parametrize the test using values that contained nested pytest.param objects,
+        # then we unpack them values and merge marks.
         newcalls = []
         for callspec in self._calls or [CallSpec2()]:
             for param_index, (param_id, param_set) in enumerate(zip(ids, parameters)):
+                normalized_values = []
```
Comment (nicoddemus) [inline src/_pytest/python.py:1134]: I would prefer if we move this to a free function and add specific unittests for it.

**Review by nicoddemus — CHANGES_REQUESTED** [review]
I left a few comments/suggestions, however I'm -0.5 on the whole idea: I understand the problem that it tries to solve, but having this done automatically by the core feels a bit too magical to me.

**Comment by vgorkavenko** [comment]
@nicoddemus Maybe we can add some argument to `parametrize` or `pytest.param` (`squash=True`, for example)? To make it less magical

**Comment by nicoddemus** [comment]
Interesting, but if we want to make it less magical/more explicit, how about adding a function which merges the parameters? For example:

```python
@pytest.mark.parametrize(
    "a,b,c", [
        pytest.merge_params(
            pytest.param(1, id="one", marks=pytest.mark.one),
            pytest.param(2, id="two", marks=pytest.mark.two),
            pytest.param(3, id="three", marks=pytest.mark.three),
            marks=pytest.mark.full
        )
    ]
)
```

This explicit and doesn't require us to change the internals.

**Comment by RonnyPfannschmidt** [comment]
I would propose a opt on/out bool kwonly parameter to pytest.param

Alternatively some kind of builder pattern

**Comment by nicoddemus** [comment]
> Alternatively some kind of builder pattern

But still in pytest itself? My impression is that this is a bit exotic use case (but I don't have any data to back that up), better made available through a plugin.

**Comment by obestwalter** [comment]
@vgorkavenko would you like to pick this up again and finish it up? Otherwise it seems more sensible to close for now and maybe pick the topic up later.

**Pull request closed without merging**